File: src/test/java/com/blackduck/integration/detect/configuration/DetectConfigurationFactoryTests.java
package com.blackduck.integration.detect.configuration;

import static com.blackduck.integration.detect.configuration.DetectConfigurationFactoryTestUtils.factoryOf;
import static com.blackduck.integration.detect.configuration.DetectConfigurationFactoryTestUtils.spyFactoryOf;
import static com.blackduck.integration.detect.configuration.DetectConfigurationFactoryTestUtils.scanSettingsFactoryOf;

import java.util.List;
import java.util.Optional;
import java.util.Collections;
import java.util.Map;
import java.util.HashMap;

import com.blackduck.integration.detect.configuration.connection.BlackDuckConnectionDetails;
import com.blackduck.integration.detect.configuration.enumeration.BlackduckScanMode;
import com.blackduck.integration.detect.configuration.enumeration.DefaultDetectorSearchExcludedDirectories;
import com.blackduck.integration.detect.configuration.enumeration.DetectTool;
import com.blackduck.integration.detect.configuration.enumeration.RapidCompareMode;
import com.blackduck.integration.detect.tool.signaturescanner.BlackDuckSignatureScannerOptions;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import com.blackduck.integration.common.util.Bdo;
import com.synopsys.integration.rest.credentials.Credentials;

public class DetectConfigurationFactoryTests {

    //#region Proxy
    @Test
    public void proxyUsesCredentials() throws DetectUserFriendlyException {
        DetectConfigurationFactory factory = factoryOf(
            Pair.of(DetectProperties.BLACKDUCK_PROXY_HOST, "host"),
            Pair.of(DetectProperties.BLACKDUCK_PROXY_PORT, "20"),
            Pair.of(DetectProperties.BLACKDUCK_PROXY_USERNAME, "username"),
            Pair.of(DetectProperties.BLACKDUCK_PROXY_PASSWORD, "password")
        );
        Bdo<Credentials> result = Bdo.of(factory.createBlackDuckProxyInfo().getProxyCredentials());

        Assertions.assertEquals(Optional.of("username"), result.flatMap(Credentials::getUsername).toOptional());
        Assertions.assertEquals(Optional.of("password"), result.flatMap(Credentials::getPassword).toOptional());
    }
    //#endregion Proxy

    //#region Parallel Processors
    @Test
    public void parallelProcessorsDefaultsToOne() {
        // Using the property default is the safe choice. See IDETECT-1970 - JM
        DetectConfigurationFactory factory = spyFactoryOf();
        Integer defaultValue = DetectProperties.DETECT_PARALLEL_PROCESSORS.getDefaultValue();

        Assertions.assertEquals(defaultValue.intValue(), factory.findParallelProcessors());
        Mockito.verify(factory, Mockito.never()).findRuntimeProcessors();
    }

    @Test
    public void parallelProcessorsPrefersProperty() {
        DetectConfigurationFactory factory = factoryOf(Pair.of(DetectProperties.DETECT_PARALLEL_PROCESSORS, "3"));

        Assertions.assertEquals(3, factory.findParallelProcessors());
    }

    //#endregion Parallel Processors

    @Test
    public void testDefaultSignatureScannerExcludedDirectories() {
        DetectConfigurationFactory factory = factoryOf();

        List<String> actualExcludedDirectories = factory.collectDetectorSearchDirectoryExclusions();

        List<String> defaultExcludedDirectories = DefaultDetectorSearchExcludedDirectories.getDirectoryNames();

        Assertions.assertEquals(defaultExcludedDirectories, actualExcludedDirectories);
    }

    @Test
    public void testIsStatelessIsEnabled() {
        DetectConfigurationFactory factory = factoryOf(
                Pair.of(DetectProperties.DETECT_TOOLS, DetectTool.SIGNATURE_SCAN.toString()),
                Pair.of(DetectProperties.DETECT_BLACKDUCK_SCAN_MODE, BlackduckScanMode.STATELESS.toString()));
        
        BlackDuckSignatureScannerOptions blackDuckSignatureScannerOptions = factory.createBlackDuckSignatureScannerOptions();

        Assertions.assertTrue(blackDuckSignatureScannerOptions.getIsStateless());
    }
    
    @Test
    public void testNoPersistenceModeSpecified() {
        DetectConfigurationFactory factory = factoryOf(
                Pair.of(DetectProperties.DETECT_BLACKDUCK_RAPID_COMPARE_MODE, RapidCompareMode.BOM_COMPARE_STRICT.toString()));
        
        BlackDuckSignatureScannerOptions blackDuckSignatureScannerOptions = factory.createBlackDuckSignatureScannerOptions();

        Assertions.assertTrue(RapidCompareMode.BOM_COMPARE_STRICT.equals(blackDuckSignatureScannerOptions.getBomCompareMode()));
    }

    @Test
    public void testGetContainerScanFilePathIfUrlProvided() {
        String imageUrl = "https://artifactory.container.com/image.tar";
        DetectConfigurationFactory factory = factoryOf(
            Pair.of(DetectProperties.DETECT_TOOLS, DetectTool.CONTAINER_SCAN.toString()),
            Pair.of(DetectProperties.DETECT_CONTAINER_SCAN_FILE, imageUrl));
        Optional<String> containerScanFilePath = factory.getContainerScanFilePath();

        Assertions.assertTrue(containerScanFilePath.isPresent());
        containerScanFilePath.ifPresent(s -> Assertions.assertEquals(s, imageUrl));
        Assertions.assertFalse(containerScanFilePath.toString().startsWith("/"));
        Assertions.assertFalse(containerScanFilePath.toString().endsWith("/"));
    }

    @Test
    public void testGetContainerScanFilePathIfLocalPathProvided() {
        String imageFilePath = "src/test/resources/tool/container.scan/testImage.tar";
        DetectConfigurationFactory factory = factoryOf(
            Pair.of(DetectProperties.DETECT_TOOLS, DetectTool.CONTAINER_SCAN.toString()),
            Pair.of(DetectProperties.DETECT_CONTAINER_SCAN_FILE, imageFilePath));
        Optional<String> containerScanFilePath = factory.getContainerScanFilePath();

        Assertions.assertTrue(containerScanFilePath.isPresent());
        containerScanFilePath.ifPresent(s -> Assertions.assertEquals(s, imageFilePath));
        Assertions.assertFalse(containerScanFilePath.toString().startsWith("http"));
        Assertions.assertFalse(containerScanFilePath.toString().startsWith("/"));
        Assertions.assertFalse(containerScanFilePath.toString().endsWith("/"));
    }

    public void testScanSettingsPropertyIfProvided() throws DetectUserFriendlyException {
        String blackduckUrl = "https://testblackduckurl.com";
        String scaasScanPath = "User/detectuser/scaasFile.txt";
        DetectConfigurationFactory factory = scanSettingsFactoryOf(Collections.emptyMap(), Pair.of(DetectProperties.BLACKDUCK_URL, blackduckUrl), Pair.of(DetectProperties.DETECT_SCAAAS_SCAN_PATH, scaasScanPath));
        BlackDuckConnectionDetails blackDuckConnectionDetails = factory.createBlackDuckConnectionDetails();

        Optional<String> blackduckUrlString = blackDuckConnectionDetails.getBlackDuckUrl();

        Assertions.assertTrue(blackduckUrlString.isPresent());
        blackduckUrlString.ifPresent(str -> Assertions.assertEquals(str, blackduckUrl));

        Optional<String> scaasFilePathString = factory.getScaaasFilePath();

        Assertions.assertTrue(scaasFilePathString.isPresent());
        scaasFilePathString.ifPresent(str -> Assertions.assertEquals(str, scaasScanPath));
    }

    public void testScanSettingsPropertyIfUserProvided() throws DetectUserFriendlyException {
        String blackduckUrl = "https://testblackduckurl.com";
        String blackduckUserUrl = "User/detectuser/scaasFile.txt";
        Map<String, String> propertyMap = new HashMap<>();
        propertyMap.put("blackduck.url", blackduckUserUrl);
        DetectConfigurationFactory factory = scanSettingsFactoryOf(propertyMap, Pair.of(DetectProperties.BLACKDUCK_URL, blackduckUrl));
        BlackDuckConnectionDetails blackDuckConnectionDetails = factory.createBlackDuckConnectionDetails();

        Optional<String> blackduckUrlString = blackDuckConnectionDetails.getBlackDuckUrl();

        Assertions.assertTrue(blackduckUrlString.isPresent());
        blackduckUrlString.ifPresent(str -> Assertions.assertEquals(str, blackduckUserUrl));
    }
}

File: src/test/java/com/synopsys/integration/detect/DetectConfigurationFactoryTests.kt
package com.synopsys.integration.detect

import com.synopsys.integration.blackduck.codelocation.signaturescanner.command.SnippetMatching
import com.synopsys.integration.configuration.config.PropertyConfiguration
import com.synopsys.integration.configuration.help.PropertyConfigurationHelpContext
import com.synopsys.integration.configuration.property.Property
import com.synopsys.integration.configuration.property.types.path.SimplePathResolver
import com.synopsys.integration.configuration.source.MapPropertySource
import com.synopsys.integration.configuration.util.configOf
import com.synopsys.integration.detect.configuration.DetectConfigurationFactory
import com.synopsys.integration.detect.configuration.DetectProperties
import org.junit.Test
import org.junit.jupiter.api.Assertions
import org.mockito.Mockito

class DetectConfigurationFactoryTests {
    @Test
    fun testFallsbackToRuntime() {
        val factory = spyFactoryOf()
        val result = factory.findParallelProcessors()
        Mockito.verify(factory).findAvailableProcessors()
    }

    @Test
    fun testsParallelChosen() {
        val factory = factoryOf(DetectProperties.DETECT_PARALLEL_PROCESSORS to "3")
        Assertions.assertEquals(3, factory.findParallelProcessors())
    }

    @Test
    fun testsPrefersNewer() {
        val factory = factoryOf(
                DetectProperties.DETECT_PARALLEL_PROCESSORS to "5",
                DetectProperties.DETECT_BLACKDUCK_SIGNATURE_SCANNER_PARALLEL_PROCESSORS to "4"
        )
        Assertions.assertEquals(5, factory.findParallelProcessors())
    }

    @Test
    fun testsWillUseOld() {
        val factory = factoryOf(
                DetectProperties.DETECT_BLACKDUCK_SIGNATURE_SCANNER_PARALLEL_PROCESSORS to "5"
        )
        Assertions.assertEquals(5, factory.findParallelProcessors())
    }

    @Test
    fun testDeprecatedSnippingTrueUsesSnippetMatchingMode() {
        val factory = factoryOf(
                DetectProperties.DETECT_BLACKDUCK_SIGNATURE_SCANNER_SNIPPET_MODE to "true"
        )
        Assertions.assertEquals(SnippetMatching.SNIPPET_MATCHING, factory.findSnippetMatching())
    }

    @Test
    fun testsNewSnippetModePreferred() {
        val factory = factoryOf(
                DetectProperties.DETECT_BLACKDUCK_SIGNATURE_SCANNER_SNIPPET_MODE to "true",
                DetectProperties.DETECT_BLACKDUCK_SIGNATURE_SCANNER_SNIPPET_MATCHING to SnippetMatching.FULL_SNIPPET_MATCHING_ONLY.name
        )
        Assertions.assertEquals(SnippetMatching.FULL_SNIPPET_MATCHING_ONLY, factory.findSnippetMatching())
    }


    private fun spyFactoryOf(vararg properties: Pair<Property, String>): DetectConfigurationFactory {
        return Mockito.spy(factoryOf(*properties));
    }

    private fun factoryOf(vararg properties: Pair<Property, String>): DetectConfigurationFactory {
        val configuration = configOf(DetectProperties.BLACKDUCK_OFFLINE_MODE.key to "unknown");
        val helpContext = PropertyConfigurationHelpContext(configuration)
        Assertions.assertEquals(1, helpContext.findPropertyParseErrors(DetectProperties.properties).size)

        val propertyMap = properties.toMap().mapKeys { it.key.key }
        val inMemoryPropertySource = MapPropertySource("test", propertyMap)
        val propertyConfiguration = PropertyConfiguration(listOf(inMemoryPropertySource))
        return DetectConfigurationFactory(propertyConfiguration, SimplePathResolver())
    }
}